# 문서 군집화

비슷한 텍스트 구성의 문서를 군집화하는것 (비지도학습 기반)

Opinion Review 데이터를 이용해 문서 군집화를 해보자

In [2]:
import pandas as pd
import glob ,os

# 아래는 제 컴퓨터에서 압축 파일을 풀어 놓은 디렉토리이니, 여러분의 디렉토리를 설정해 주십시요  
path = r'C:\Users\lmj\Desktop\데이터세트\OpinosisDataset1.0\OpinosisDataset1.0\topics'                     
# path로 지정한 디렉토리 밑에 있는 모든 .data 파일들의 파일명을 리스트로 취합
all_files = glob.glob(os.path.join(path, "*.data"))    
filename_list = []
opinion_text = []

# 개별 파일들의 파일명은 filename_list 리스트로 취합, 
# 개별 파일들의 파일내용은 DataFrame로딩 후 다시 string으로 변환하여 opinion_text 리스트로 취합 
for file_ in all_files:
    # 개별 파일을 읽어서 DataFrame으로 생성 
    df = pd.read_table(file_,index_col=None, header=0,encoding='latin1')
    
    # 절대경로로 주어진 file 명을 가공. 만일 Linux에서 수행시에는 아래 \\를 / 변경. 맨 마지막 .data 확장자도 제거
    filename_ = file_.split('\\')[-1]
    filename = filename_.split('.')[0]

    #파일명 리스트와 파일내용 리스트에 파일명과 파일 내용을 추가. 
    filename_list.append(filename)
    opinion_text.append(df.to_string())

# 파일명 리스트와 파일내용 리스트를  DataFrame으로 생성
document_df = pd.DataFrame({'filename':filename_list, 'opinion_text':opinion_text})
document_df.head()

filename  \
0   accuracy_garmin_nuvi_255W_gps   
1  bathroom_bestwestern_hotel_sfo   
2      battery-life_amazon_kindle   
3      battery-life_ipod_nano_8gb   
4     battery-life_netbook_1005ha   

                                        opinion_text  
0                                                ...  
1                                                ...  
2                                                ...  
3                                                ...  
4                                                ...

압축을 해제한 경로 내의 모든 파일에 대해 for문으로 반복하면서 파일명과 파일내용을 리스트에 추가한 뒤 DataFrame으로 변환해주었다.

이 문서를 TF-IDF 형태로 피쳐 벡터화를 해줄것이다.

토큰화는 Lemmatizaton으로 / n-gram은 (1,2)로 / min_df와 max_df범위를 설정해 피처의 개수를 제한하였다.

In [3]:
from nltk.stem import WordNetLemmatizer
import nltk
import string

remove_punct_dict = dict((ord(punct), None) for punct in string.punctuation)
lemmar = WordNetLemmatizer()

# 토큰화
def LemTokens(tokens):
    return [lemmar.lemmatize(token) for token in tokens]
#정규화
def LemNormalize(text):
    return LemTokens(nltk.word_tokenize(text.lower().translate(remove_punct_dict)))

from sklearn.feature_extraction.text import TfidfVectorizer

# 벡터화
tfidf_vect = TfidfVectorizer(tokenizer=LemNormalize, stop_words='english' , \
                             ngram_range=(1,2), min_df=0.05, max_df=0.85 )

#opinion_text 컬럼값으로 feature vectorization 수행
feature_vect = tfidf_vect.fit_transform(document_df['opinion_text'])

C:\Users\lmj\anaconda3\lib\site-packages\sklearn\feature_extraction\text.py:396: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['ha', 'le', 'u', 'wa'] not in stop_words.
  warnings.warn(


군집화 방식은 K-평균을 이용하였다. 그리고 군집의 Label값과 할당된 데이터세트의 좌표값을 구하였다.

In [4]:
from sklearn.cluster import KMeans

# 5개 집합으로 군집화 수행. 예제를 위해 동일한 클러스터링 결과 도출용 random_state=0 
km_cluster = KMeans(n_clusters=5, max_iter=10000, random_state=0)
km_cluster.fit(feature_vect)
cluster_label = km_cluster.labels_
cluster_centers = km_cluster.cluster_centers_

군집이 주제별로 유사하게 잘 구성되었는지 살펴보자

In [5]:
document_df['cluster_label'] = cluster_label
document_df.head()

filename  \
0   accuracy_garmin_nuvi_255W_gps   
1  bathroom_bestwestern_hotel_sfo   
2      battery-life_amazon_kindle   
3      battery-life_ipod_nano_8gb   
4     battery-life_netbook_1005ha   

                                        opinion_text  cluster_label  
0                                                ...              2  
1                                                ...              0  
2                                                ...              1  
3                                                ...              1  
4                                                ...              1

In [6]:
document_df[document_df['cluster_label']==0].sort_values(by='filename')

filename  \
1   bathroom_bestwestern_hotel_sfo   
32         room_holiday_inn_london   
30     rooms_bestwestern_hotel_sfo   
31         rooms_swissotel_chicago   

                                         opinion_text  cluster_label  
1                                                 ...              0  
32                                                ...              0  
30                                                ...              0  
31                                                ...              0

0번 군집은 호텔에 관련된 군집

In [7]:
document_df[document_df['cluster_label']==1].sort_values(by='filename')

filename  \
2    battery-life_amazon_kindle   
3    battery-life_ipod_nano_8gb   
4   battery-life_netbook_1005ha   
19      keyboard_netbook_1005ha   
26   performance_netbook_1005ha   
41     size_asus_netbook_1005ha   
42          sound_ipod_nano_8gb   
44               speed_windows7   

                                         opinion_text  cluster_label  
2                                                 ...              1  
3                                                 ...              1  
4                                                 ...              1  
19                                                ...              1  
26                                                ...              1  
41                                                ...              1  
42      headphone jack i got a clear case for it a...              1  
44                                                ...              1

1번 군집은 포터블 기기에 관한 군집

In [8]:
document_df[document_df['cluster_label']==2].sort_values(by='filename')

filename  \
0     accuracy_garmin_nuvi_255W_gps   
5             buttons_amazon_kindle   
8   directions_garmin_nuvi_255W_gps   
9      display_garmin_nuvi_255W_gps   
10    eyesight-issues_amazon_kindle   
11                features_windows7   
12              fonts_amazon_kindle   
23         navigation_amazon_kindle   
33   satellite_garmin_nuvi_255W_gps   
34      screen_garmin_nuvi_255W_gps   
35             screen_ipod_nano_8gb   
36            screen_netbook_1005ha   
43       speed_garmin_nuvi_255W_gps   
48     updates_garmin_nuvi_255W_gps   
49              video_ipod_nano_8gb   
50       voice_garmin_nuvi_255W_gps   

                                         opinion_text  cluster_label  
0                                                 ...              2  
5                                                 ...              2  
8                                                 ...              2  
9                                                 ...              2  
10                                                ...              2  
11                                                ...              2  
12                                                ...              2  
23                                                ...              2  
33                                                ...              2  
34                                                ...              2  
35                                                ...              2  
36                                                ...              2  
43                                                ...              2  
48                                                ...              2  
49                                                ...              2  
50                                                ...              2

2번 군집은 차량용 네비게이션 군집

이하생략

전반적인 군집을 살펴보았을 때 군집 갯수가 많다고 판단되어 군집 개수를 3개로 낮추어보았다

In [12]:
from sklearn.cluster import KMeans

# 3개의 집합으로 군집화 
km_cluster = KMeans(n_clusters=3, max_iter=10000, random_state=0)
km_cluster.fit(feature_vect)
cluster_label = km_cluster.labels_


# 소속 클러스터를 cluster_label 컬럼으로 할당하고 cluster_label 값으로 정렬
document_df['cluster_label'] = cluster_label
document_df.sort_values(by='cluster_label')

filename  \
0     accuracy_garmin_nuvi_255W_gps   
48     updates_garmin_nuvi_255W_gps   
44                   speed_windows7   
43       speed_garmin_nuvi_255W_gps   
42              sound_ipod_nano_8gb   
41         size_asus_netbook_1005ha   
36            screen_netbook_1005ha   
35             screen_ipod_nano_8gb   
34      screen_garmin_nuvi_255W_gps   
33   satellite_garmin_nuvi_255W_gps   
27              price_amazon_kindle   
26       performance_netbook_1005ha   
49              video_ipod_nano_8gb   
23         navigation_amazon_kindle   
19          keyboard_netbook_1005ha   
50       voice_garmin_nuvi_255W_gps   
9      display_garmin_nuvi_255W_gps   
2        battery-life_amazon_kindle   
3        battery-life_ipod_nano_8gb   
4       battery-life_netbook_1005ha   
5             buttons_amazon_kindle   
12              fonts_amazon_kindle   
11                features_windows7   
10    eyesight-issues_amazon_kindle   
8   directions_garmin_nuvi_255W_gps   
47   transmission_toyota_camry_2007   
37          seats_honda_accord_2008   
6         comfort_honda_accord_2008   
7         comfort_toyota_camry_2007   
16    gas_mileage_toyota_camry_2007   
25    performance_honda_accord_2008   
17       interior_honda_accord_2008   
18       interior_toyota_camry_2007   
22        mileage_honda_accord_2008   
29        quality_toyota_camry_2007   
1    bathroom_bestwestern_hotel_sfo   
46          staff_swissotel_chicago   
45      staff_bestwestern_hotel_sfo   
14           food_swissotel_chicago   
20   location_bestwestern_hotel_sfo   
21      location_holiday_inn_london   
30      rooms_bestwestern_hotel_sfo   
38    service_bestwestern_hotel_sfo   
13          food_holiday_inn_london   
24    parking_bestwestern_hotel_sfo   
28         price_holiday_inn_london   
15       free_bestwestern_hotel_sfo   
32          room_holiday_inn_london   
31          rooms_swissotel_chicago   
39       service_holiday_inn_london   
40  service_swissotel_hotel_chicago   

                                         opinion_text  cluster_label  
0                                                 ...              0  
48                                                ...              0  
44                                                ...              0  
43                                                ...              0  
42      headphone jack i got a clear case for it a...              0  
41                                                ...              0  
36                                                ...              0  
35                                                ...              0  
34                                                ...              0  
33                                                ...              0  
27                                                ...              0  
26                                                ...              0  
49                                                ...              0  
23                                                ...              0  
19                                                ...              0  
50                                                ...              0  
9                                                 ...              0  
2                                                 ...              0  
3                                                 ...              0  
4                                                 ...              0  
5                                                 ...              0  
12                                                ...              0  
11                                                ...              0  
10                                                ...              0  
8                                                 ...              0  
47                                                ...              1  
37                                                ...              1  
6         

0번 군집은 포터블 전자기기/ 1번 군집은 호텔 리뷰/ 2번 군집은 자동차 리뷰

---

# 핵심 단어 추출

군집에 속한 문서는 핵심단어를 주축으로 군집화되어있다. 그래서 군집을 구성하는 핵심 단어를 추출해보자

k평균 군집화에서 객체는 cluster_centers_ 속성을 통해 군집의 중심을 기준으로 가까이 위치했는지 알 수 있다.

In [13]:
cluster_centers = km_cluster.cluster_centers_
print('cluster_centers shape :',cluster_centers.shape)
print(cluster_centers)

cluster_centers shape : (3, 4611)
[[0.01005322 0.         0.         ... 0.00706287 0.         0.        ]
 [0.         0.00092551 0.         ... 0.         0.         0.        ]
 [0.         0.00099499 0.00174637 ... 0.         0.00183397 0.00144581]]


각 배열 내의 값은 군집 내의 상대 위치를 표현한 일종의 좌표값이다.

(3,4611)은 3개의 군집, 4611개의 word피처로 구성되어있다는 뜻이다.

이를 이용하여 핵심 단어를 추출해보자

In [14]:
# 군집별 top n 핵심단어, 그 단어의 중심 위치 상대값, 대상 파일명들을 반환함. 
def get_cluster_details(cluster_model, cluster_data, feature_names, clusters_num, top_n_features=10):
    cluster_details = {}
    
    # cluster_centers array 의 값이 큰 순으로 정렬된 index 값을 반환
    # 군집 중심점(centroid)별 할당된 word 피처들의 거리값이 큰 순으로 값을 구하기 위함.  
    centroid_feature_ordered_ind = cluster_model.cluster_centers_.argsort()[:,::-1]
    
    #개별 군집별로 iteration하면서 핵심단어, 그 단어의 중심 위치 상대값, 대상 파일명 입력
    for cluster_num in range(clusters_num):
        # 개별 군집별 정보를 담을 데이터 초기화. 
        cluster_details[cluster_num] = {}
        cluster_details[cluster_num]['cluster'] = cluster_num
        
        # cluster_centers_.argsort()[:,::-1] 로 구한 index 를 이용하여 top n 피처 단어를 구함. 
        top_feature_indexes = centroid_feature_ordered_ind[cluster_num, :top_n_features]
        top_features = [ feature_names[ind] for ind in top_feature_indexes ]
        
        # top_feature_indexes를 이용해 해당 피처 단어의 중심 위치 상댓값 구함 
        top_feature_values = cluster_model.cluster_centers_[cluster_num, top_feature_indexes].tolist()
        
        # cluster_details 딕셔너리 객체에 개별 군집별 핵심 단어와 중심위치 상대값, 그리고 해당 파일명 입력
        cluster_details[cluster_num]['top_features'] = top_features
        cluster_details[cluster_num]['top_features_value'] = top_feature_values
        filenames = cluster_data[cluster_data['cluster_label'] == cluster_num]['filename']
        filenames = filenames.values.tolist()
        cluster_details[cluster_num]['filenames'] = filenames
        
    return cluster_details

In [ ]:
def print_cluster_details(cluster_details):
    for cluster_num, cluster_detail in cluster_details.items():
        print('####### Cluster {0}'.format(cluster_num))
        print('Top features:', cluster_detail['top_features'])
        print('Reviews 파일명 :',cluster_detail['filenames'][:7])
        print('==================================================')

In [16]:
feature_names = tfidf_vect.get_feature_names()

cluster_details = get_cluster_details(cluster_model=km_cluster, cluster_data=document_df,\
                                  feature_names=feature_names, clusters_num=3, top_n_features=10 )
print_cluster_details(cluster_details)

####### Cluster 0
Top features: ['screen', 'battery', 'keyboard', 'battery life', 'life', 'kindle', 'direction', 'video', 'size', 'voice']
Reviews 파일명 : ['accuracy_garmin_nuvi_255W_gps', 'battery-life_amazon_kindle', 'battery-life_ipod_nano_8gb', 'battery-life_netbook_1005ha', 'buttons_amazon_kindle', 'directions_garmin_nuvi_255W_gps', 'display_garmin_nuvi_255W_gps']
####### Cluster 1
Top features: ['interior', 'seat', 'mileage', 'comfortable', 'gas', 'gas mileage', 'transmission', 'car', 'performance', 'quality']
Reviews 파일명 : ['comfort_honda_accord_2008', 'comfort_toyota_camry_2007', 'gas_mileage_toyota_camry_2007', 'interior_honda_accord_2008', 'interior_toyota_camry_2007', 'mileage_honda_accord_2008', 'performance_honda_accord_2008']
####### Cluster 2
Top features: ['room', 'hotel', 'service', 'staff', 'food', 'location', 'bathroom', 'clean', 'price', 'parking']
Reviews 파일명 : ['bathroom_bestwestern_hotel_sfo', 'food_holiday_inn_london', 'food_swissotel_chicago', 'free_bestwestern_h

C:\Users\lmj\anaconda3\lib\site-packages\sklearn\utils\deprecation.py:87: FutureWarning: Function get_feature_names is deprecated; get_feature_names is deprecated in 1.0 and will be removed in 1.2. Please use get_feature_names_out instead.
  warnings.warn(msg, category=FutureWarning)


각 군집별로  핵심단어가 추출되었다.

0번군집은 포터블 기기/ 1번군집은 호텔/ 2번군집은 자동차 로 주요 관심사가 무었인지 알 수 있다.